src/DominantJuice.sol

```diff
@@ -0,0 +1,399 @@
+    IJBDirectory public immutable directory; // directory of terminals and controllers for projects.
+    IJBFundAccessConstraintsStore public immutable fundAccessConstraintsStore;
+    IJBSingleTokenPaymentTerminalStore3_1_1 public immutable paymentTerminalStore;
+    IJBSingleTokenPaymentTerminal public immutable paymentTerminal;
+
+    /// Main Campaign Parameters
+    struct Campaign {
+        uint256 projectId;
+        uint256 cycleTarget;
+        uint256 cycleExpiryDate;
+        uint256 minimumPledgeAmount;
+        uint256 totalRefundBonus;
+    }
+
+    struct FundingStatus {
+        uint256 totalPledged;
+        uint256 percentOfGoal;
+        bool isTargetMet;
+        bool hasCycleExpired;
+        bool hasCreatorWithdrawnAllFunds;
+    }
+
+    struct Pledgers {
+        uint256 totalAmountPledged;
+        UD60x18 totalPledgeWeight;
+        mapping(address => UD60x18) pledgerWeight;
+        mapping(address => bool) hasBeenRefunded;
+    }
+
+    /// State Variables
+    Campaign private campaign;
+    Pledgers private pledgers;
+    UD60x18 private immutable rateOfDecay = ud(0.99e18);
+
+    // Events for transparency to pledgers
+    event RefundBonusDeposited(address, uint256 indexed);
+    event PledgeMade(address, uint256);
+    event CycleHasClosed(bool indexed, bool indexed);
+    event CycleRefundBonusWithdrawal(address indexed, uint256 indexed);
+    event CreatorWithdrawal(address, uint256);
+
+    error BonusAlreadyDeposited(uint256 bonusAmount);
+    error RefundBonusNotDeposited();
+    error AmountIsBelowMinimumPledge(uint256 minAmount);
+    error InvalidPaymentEvent(address caller, uint256 projectId, uint256 value);
+    error CycleHasExpired();
+    error IncorrectProjectID();
+    error FunctionHasAlreadyBeenCalled();
+    error MustBePledger();
+    error CycleHasNotEndedYet(uint256 endTimestamp);
+    error NoRefundsForSuccessfulCycle();
+    error AlreadyWithdrawnRefund();
+    error InvalidRedemptionEvent(address caller, uint256 projectId, uint256 value);
+    error InsufficientFunds();
+
+    // Make sure the caller is a terminal of the project, and that the call is being made
+    // on behalf of an interaction with the correct project.
+    modifier terminalCheck() {
+        require(msg.value == 0, "Pledges should be made through JB website.");
+        require(
+            directory.isTerminalOf(campaign.projectId, IJBPaymentTerminal(msg.sender)),
+            "Caller must be a JB Payment Terminal."
+        );
+        _;
+    }
+
+    modifier redeemCycleCheck() {
+        // This is an insurance check, since project should have redemptions paused during 1st cycle.
+        if (!hasCycleExpired()) revert CycleHasNotEndedYet(campaign.cycleExpiryDate);
+        if (isTargetMet()) revert NoRefundsForSuccessfulCycle();
+        _;
+    }
+
+    /// @param _projectId Obtained via Juicebox after project creation.
+    constructor(
+        address payable _campaignManager,
```

Comment:
a struct for all the constructor params would be better than this many params. Anything more than 3 gets unwieldy.

Reply:
Would there be heavy overlap with the Campaign struct? How would you suggest this gets implemented?

Reply:
If they are the same, then sure, but I tend to just create an init struct eg `DominantJuiceInit` to house the variables